### DISCUSSION TITLE
Vevor level 1/2 EV Charger

### DISCUSSION THREAD

**Discussion by thegnomewizard** [arendst/Tasmota#21385] [Support request]
Hey all - this may be better filed as a simple Issue, but thought I'd start here.

Basically, though I'm a software engineer I'm fairly new to Tasmota, used it in a few blinds/etc. but this is my first time using it in a non-supported device. Don't feel confident posting a template yet, but I do have some info.

These chargers, available on Amazon for only $100 for the non-LCD screen version, contain a Tuya module, though it's not an ESP. It's a WBR2 (for which there are SDKs floating around, but that's obviously not the target of the Tasmota project), but it's compatible with the ESP-02S so I grabbed a few of those from Aliexpress, slapped Tasmota in one, and installed it. As an aside, the rest of the device seems to be quite nicely designed for the price, no shockingly (haha) bad choices on the PCB that I could see. The real manufacturer of the system/PCB is "Zopoise".

Anyway, after setting the Tuya RX/TX pins to the ones marked RX/TX on the ESP-02S itself (it looked like those were the only parts hooked up on the main PCB other than power) and setting it to 115200bps, I could indeed talk to the MCU. I got some details from this thread: https://github.com/make-all/tuya-local/issues/1757 (mostly useful as I dislike Tuya the company enough that I didn't want to connect it even once just to grab the info myself). The first post seems to be incorrect but does list the various enum states, if you google some of those you can also find other instances. For ref here's a little spreadsheet I made with the mappins to Tasmota I guessed (I included more info in the "comments" but sadly the Github import won't capture those, and also I think some of the units are probably wrong from the thread):

![image](https://github.com/arendst/Tasmota/assets/116392349/7f8256cb-61d4-4cfa-8283-80c72b5f0af2)

Either way, everything largely works, but I have a couple of questions:

1) phase_a (dpId 6) is reporting a string of bytes (I note it's described as base64 in that thread, but I can see if it was, it's already been decoded in the logs I get from Tasmota). I assume I'll just need to send these out over MQTT and decode them on my server via Python? The message contains the current readout voltage/current/etc. info.
2) power_total (dpId 9) is the current charge power, in watts, e.g. for 1kW it reports 1000. It looks like Tasmota currently assumes this field is always deciwatts - I assume there's no multiplier like there is for the temperature in dpId 24 (which is just an integer temperature reading)?
3) weirdest one of all - the MCU never sends updates. If I run TuyaSend0 or TuyaSend8, I get the current values I want to appear on the Tasmota home page/send over MQTT, but it never sends them unprompted. TuyaSend9 doesn't do anything beyond make Tasmota say it's turned on/off and save that setting to flash.

1 and 2 I can obviously solve on my server end, 3 I can fix with a rule triggering every few seconds, but I just wondered if there's anything else I can try? Obviously other devices may also come out with slightly odd readings in power/etc. (I know there's mention of some outlets also using the same binary format for phase_a) so presumably I should file issues for at least those ones?

Thanks! :)

**Comment by barbudor** [comment]
> this may be better filed as a simple Issue, but thought I'd start here

You did it correctly. Issues are for fully qualified bugs with all information for reproducing to allow debugging by the dev team.

Not getting what you expect from a brand new device, especially one that has been brain-converted is not an issue but a call for help.
Evenually it may be come also a feature-request later if some missing featuers are identified, but that would still be here in Discussion


Back to your problem, you are not clear on what dpId/fnId mapping you tried and haven't provided any logs.

A properly mapped dpId/fnId should be decoded by Tasmota and should be presented in hte SENSOR message

On dpId 6, if it is a combined voltage/current/power, the proper fnId is 36 and should be decoded by Tasmota (if properly configured) - obviously doc need to be updated for those missing fnId that can be found in the code: https://github.com/arendst/Tasmota/blob/2f3663b8aa838e25272c567b8e76ff904cc74cf0/tasmota/include/tasmota.h#L551-L570)

On dpId9, if it is power, fnId should 31 and indeed Tasmota is considering it as deciWatt as [documented](https://tasmota.github.io/docs/TuyaMCU/#fnid)

On your 3rd point, TuyaMCU not sending may any udpdate could be that it doesn't consider that it is connected to it's cloud
For Tasmota to say to the MCU that it is connected, if MQTT is enabled, you must be connected to the broker.
Having MQTT enabled but not properly configured would generally lead to MCU not sending updates, and in some case even resetting regularly the ESP until it gets the mesage it wants


Definitively providing logs at level 3 would be helpfull to understand the situation



Alternatively there is a TuyaMCU Bridge mode (using GPIOs "TuyaMCUBr Tx/Rx" which will transfer any Tuya message over MQTT to your backend for processing htere. So you will have to fully handle the protocol on your side instead of letting Tasmota doing it.
I never used that mode

**Comment by thegnomewizard** [comment]
Thanks for the pointer! Setting fnId 31 to dpId 9 does indeed populate the voltage, power and current (I did also need to disable dpId 9 as that appears to still override it with the divided number, but that's fine) so that solves issues 1 and 2. I had actually had a quick look through the code but it's always hard to find something you're not even sure if it's there, or what it might be called.

MQTT was already working obviously, got many other Tasmota or alike devices connected to the system already. I was also able to extract the other random info (like the charge state) for Prometheus by turning on the TuyaMCU to MQTT bridge.

<img width="398" alt="Screenshot 2024-05-08 at 9 52 44 pm" src="https://github.com/arendst/Tasmota/assets/116392349/452b0b5a-7101-41c4-9ae4-e3c818e7f88b">

Basically everything works now, just so long as I set up a timer to manually run TuyaSend0, my issue 3 above. Here's a log for what happens after I start up the ESP-02S:

```
00:00:00.001 HDW: ESP8285N08
00:00:00.054 CFG: Loaded from flash at F5, Count 143
00:00:00.059 QPC: Count 1
00:00:00.061 CFG: CR 384/699, Busy 0
00:00:00.065 TYA: Active=1
00:00:00.069 ROT: Mode 1
00:00:00.070 NRG: Init driver 32
00:00:00.071 SNS: Hardware Serial
00:00:00.072 TYA: Request MCU configuration at 115200 bps
00:00:00.073 SRC: Restart
00:00:00.074 TYA: Send "55aa0006000512010001011f"
00:00:00.096 Project tasmota - VevorCharger Version 13.4.0(tasmota)-2_7_6(2024-02-14T16:13:56)
00:00:00.199 WIF: Checking connection...
00:00:00.199 WIF: Attempting connection...
00:00:01.001 WIF: Connecting to AP1 gvhs Channel 8 BSSId 48:A9:8A:5B:34:33 in mode 11n as vevorcharge1-2049...
00:00:01.003 TYA: Set WiFi LED 2 (-1)
00:00:01.005 TYA: Send "55aa000300010205"
00:00:01.066 DMP: 55 AA 03 03 00 00 05
00:00:01.068 RSL: RESULT = {"TuyaReceived":{"Data":"55AA0303000005","Cmnd":3}}
00:00:01.069 TYA: RX WiFi LED set ACK
00:00:01.870 WIF: Checking connection...
00:00:01.871 WIF: Attempting connection...
00:00:02.824 WIF: Checking connection...
00:00:02.825 WIF: Attempting connection...
00:00:03.827 WIF: Checking connection...
00:00:03.828 WIF: Connected
00:00:04.079 HTP: Web server active on vevorcharge1-2049 with IP address 10.116.205.154
00:00:04.234 HTP: Console
00:00:04.289 WIF: Sending Gratuitous ARP
00:00:04.291 NTP: Sync time...
00:00:04.306 WIF: DNS resolved '2.pool.ntp.org' (44.4.53.4) in 13 ms
00:00:04.410 RTC: UTC 2024-05-09T04:46:08Z, DST 2024-03-31T02:00:00, STD 2024-10-27T03:00:00
05:46:08.000 RTC: Synced by NTP
05:46:08.002 TYA: Set WiFi LED 3 (0)
05:46:08.004 TYA: Send "55aa000300010306"
05:46:09.316 MQT: tele/vevorcharge1/RESULT = {"TuyaReceived":{"Data":"55AA03070008040200040000000C27","Cmnd":7,"CmndData":"040200040000000C","DpType2Id4":12,"4":{"DpId":4,"DpIdType":2,"DpIdData":"0000000C"}}}
05:46:09.319 TYA: fnId=99 is set for dpId=4
05:46:09.321 TYA: RX value 12 from dpId 4 
05:46:09.326 DMP: 55 AA 03 07 00 0C 06 00 00 08 04 A4 00 2C 88 00 05 4A CE
05:46:09.331 MQT: tele/vevorcharge1/RESULT = {"TuyaReceived":{"Data":"55AA0307000C0600000804A4002C8800054ACE","Cmnd":7,"CmndData":"0600000804A4002C8800054A","DpType0Id6":"0x04A4002C8800054A","6":{"DpId":6,"DpIdType":0,"DpIdData":"04A4002C8800054A"}}}
05:46:09.336 TYA: fnId=36 is set for dpId=6
05:46:09.337 TYA: Rx ID=6 Voltage=1188
05:46:09.339 TYA: Rx ID=6 Current=11400
05:46:09.341 TYA: Rx ID=6 Active_Power=1354
05:46:09.344 DMP: 55 AA 03 07 00 08 09 02 00 04 00 00 05 4A 6F
05:46:09.349 MQT: tele/vevorcharge1/RESULT = {"TuyaReceived":{"Data":"55AA03070008090200040000054A6F","Cmnd":7,"CmndData":"090200040000054A","DpType2Id9":1354,"9":{"DpId":9,"DpIdType":2,"DpIdData":"0000054A"}}}
05:46:09.352 TYA: fnId=0 is set for dpId=9
05:46:09.354 TYA: RX value 1354 from dpId 9 
05:46:09.358 DMP: 55 AA 03 07 00 06 0A 05 00 02 00 00 20
05:46:09.362 MQT: tele/vevorcharge1/RESULT = {"TuyaReceived":{"Data":"55AA030700060A050002000020","Cmnd":7,"CmndData":"0A0500020000","DpType5Id10":"0x0000","10":{"DpId":10,"DpIdType":5,"DpIdData":"0000"}}}
05:46:09.366 TYA: fnId=0 is set for dpId=10
05:46:09.371 DMP: 55 AA 03 07 00 05 0D 04 00 01 05 25
05:46:09.377 MQT: tele/vevorcharge1/RESULT = {"TuyaReceived":{"Data":"55AA030700050D0400010525","Cmnd":7,"CmndData":"0D04000105","DpType4Id13":5,"13":{"DpId":13,"DpIdType":4,"DpIdData":"05"}}}
05:46:09.380 TYA: fnId=62 is set for dpId=13
05:46:09.382 SRC: Switch
05:46:09.384 CMD: Grp 0, Cmd 'TUYAENUM', Idx 2, Len 1, Pld 5, Data '5'
05:46:09.391 MQT: stat/vevorcharge1/RESULT = {"TuyaEnum2":5}
05:46:09.395 DMP: 55 AA 03 07 00 0A 17 03 00 06 56 31 2E 30 2E 37 7D
05:46:09.399 MQT: tele/vevorcharge1/RESULT = {"TuyaReceived":{"Data":"55AA0307000A1703000656312E302E377D","Cmnd":7,"CmndData":"1703000656312E302E37","DpType3Id23":"V1.0.7","23":{"DpId":23,"DpIdType":3,"DpIdData":"56312E302E37","Type3Data":"V1.0.7"}}}
05:46:09.403 TYA: fnId=0 is set for dpId=23
05:46:09.407 DMP: 55 AA 03 07 00 08 18 02 00 04 00 00 00 30 5F
05:46:09.412 MQT: tele/vevorcharge1/RESULT = {"TuyaReceived":{"Data":"55AA0307000818020004000000305F","Cmnd":7,"CmndData":"1802000400000030","DpType2Id24":48,"24":{"DpId":24,"DpIdType":2,"DpIdData":"00000030"}}}
05:46:09.414 TYA: fnId=71 is set for dpId=24
05:46:09.417 DMP: 55 AA 03 07 00 06 21 01 00 00 00 01 32
05:46:09.421 MQT: tele/vevorcharge1/RESULT = {"TuyaReceived":{"Data":"55AA0307000621010000000132","Cmnd":7,"CmndData":"210100000001","DpType1Id33":"0x","33":{"DpId":33,"DpIdType":1,"DpIdData":""}}}
05:46:09.424 TYA: fnId=0 is set for dpId=33
05:46:10.362 QPC: Reset
05:46:10.365 TYA: Set WiFi LED 4 (0)
05:46:10.369 TYA: Send "55aa000300010407"
05:46:10.372 DMP: 55 AA 03 03 00 00 05
05:46:10.377 MQT: tele/vevorcharge1/RESULT = {"TuyaReceived":{"Data":"55AA0303000005","Cmnd":3}}
05:46:10.380 TYA: RX WiFi LED set ACK
05:46:11.227 NRG: Energy reset by invalid data
05:46:12.228 APP: Boot Count 32
05:46:12.732 CFG: Saved to flash at F4, Count 144, Bytes 4096
05:46:13.248 MQT: tele/vevorcharge1/STATE = {"Time":"2024-05-09T05:46:13","Uptime":"0T00:00:11","UptimeSec":11,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":24,"MqttCount":1,"POWER":"ON","Wifi":{"AP":1,"SSId":"gvhs","BSSId":"48:A9:8A:5B:34:33","Channel":8,"Mode":"11n","RSSI":66,"Signal":-67,"LinkCount":1,"Downtime":"0T00:00:05"}}
05:46:13.262 MQT: tele/vevorcharge1/SENSOR = {"Time":"2024-05-09T05:46:13","ENERGY":{"TotalStartTime":"2024-05-07T10:50:01","Total":4.142,"Yesterday":2.964,"Today":1.174,"Period":0,"Power":0,"ApparentPower":0,"ReactivePower":0,"Factor":0.00,"Voltage":0,"Current":0.000},"TuyaSNS":{"Temperature":48,"Timer1":null},"TempUnit":"C"}
05:46:15.240 TYA: Send "55aa00000000ff"
05:46:15.243 DMP: 55 AA 03 00 00 01 01 04
05:46:15.248 MQT: tele/vevorcharge1/RESULT = {"TuyaReceived":{"Data":"55AA030000010104","Cmnd":0,"CmndData":"01"}}
05:46:15.250 TYA: Heartbeat
05:46:18.309 MQT: tasmota/discovery/3C6105972801/config = {"ip":"10.116.205.154","dn":"VevorCharger","fn":["VevorCharger",null,null,null,null,null,null,null],"hn":"vevorcharge1-2049","mac":"3C6105972801","md":"Tuya MCU","ty":1,"if":0,"ofln":"Offline","onln":"Online","state":["OFF","ON","TOGGLE","HOLD"],"sw":"13.4.0","t":"vevorcharge1","ft":"%prefix%/%topic%/","tp":["cmnd","stat","tele"],"rl":[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"swc":[-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1],"swn":[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],"btn":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"so":{"4":0,"11":0,"13":0,"17":0,"20":0,"30":0,"68":0,"73":0,"82":0,"114":0,"117":0},"lk":0,"lt_st":0,"bat":0,"dslp":0,"sho":[],"sht":[],"ver":1} (retained)
05:46:18.323 MQT: tasmota/discovery/3C6105972801/sensors = {"sn":{"Time":"2024-05-09T05:46:18","ENERGY":{"TotalStartTime":"2024-05-07T10:50:01","Total":4.142,"Yesterday":2.964,"Today":1.174,"Power":0,"ApparentPower":0,"ReactivePower":0,"Factor":0.00,"Voltage":0,"Current":0.000},"TuyaSNS":{"Temperature":48,"Timer1":null},"TempUnit":"C"},"ver":1} (retained)
05:46:26.269 TYA: Send "55aa00000000ff"
05:46:26.272 DMP: 55 AA 03 00 00 01 01 04
05:46:26.277 MQT: tele/vevorcharge1/RESULT = {"TuyaReceived":{"Data":"55AA030000010104","Cmnd":0,"CmndData":"01"}}
05:46:26.279 TYA: Heartbeat
05:46:27.832 WIF: Checking connection...
05:46:31.007 HTP: Console
05:46:33.634 HTP: Main Menu
05:46:37.257 TYA: Send "55aa00000000ff"
05:46:37.260 DMP: 55 AA 03 00 00 01 01 04
05:46:37.268 MQT: tele/vevorcharge1/RESULT = {"TuyaReceived":{"Data":"55AA030000010104","Cmnd":0,"CmndData":"01"}}
05:46:37.271 TYA: Heartbeat
05:46:47.859 WIF: Checking connection...
05:46:48.282 TYA: Send "55aa00000000ff"
05:46:48.375 DMP: 55 AA 03 00 00 01 01 04
05:46:48.381 MQT: tele/vevorcharge1/RESULT = {"TuyaReceived":{"Data":"55AA030000010104","Cmnd":0,"CmndData":"01"}}
05:46:48.383 TYA: Heartbeat
05:46:59.244 TYA: Send "55aa00000000ff"
05:46:59.309 DMP: 55 AA 03 00 00 01 01 04
05:46:59.314 MQT: tele/vevorcharge1/RESULT = {"TuyaReceived":{"Data":"55AA030000010104","Cmnd":0,"CmndData":"01"}}
05:46:59.316 TYA: Heartbeat
05:47:02.272 TYA: Send "55aa001c000801180509052f020484"
05:47:04.227 WIF: Sending Gratuitous ARP
05:47:07.827 WIF: Checking connection...
05:47:10.233 TYA: Send "55aa00000000ff"
05:47:10.241 DMP: 55 AA 03 00 00 01 01 04
05:47:10.246 MQT: tele/vevorcharge1/RESULT = {"TuyaReceived":{"Data":"55AA030000010104","Cmnd":0,"CmndData":"01"}}
05:47:10.249 TYA: Heartbeat
05:47:21.229 TYA: Send "55aa00000000ff"
05:47:21.233 DMP: 55 AA 03 00 00 01 01 04
05:47:21.238 MQT: tele/vevorcharge1/RESULT = {"TuyaReceived":{"Data":"55AA030000010104","Cmnd":0,"CmndData":"01"}}
05:47:21.240 TYA: Heartbeat
05:47:27.861 WIF: Checking connection...
05:47:32.236 TYA: Send "55aa00000000ff"
05:47:32.255 DMP: 55 AA 03 00 00 01 01 04
05:47:32.260 MQT: tele/vevorcharge1/RESULT = {"TuyaReceived":{"Data":"55AA030000010104","Cmnd":0,"CmndData":"01"}}
05:47:32.262 TYA: Heartbeat
```

**Reply by stundenblume** [reply]
> Basically everything works now, just so long as I set up a timer to manually run TuyaSend0, my issue 3 above. Here's a log for what happens after I start up the ESP-02S:

@thegnomewizard 

can you set the charge current?

Thx, Sebastian

**Comment by thegnomewizard** [comment]
Okay, trying it again (after noticing the above log _did_ mention unrequested updates) they do very occasionally appear from the MCU automatically, but after sitting watching it for a bit, it looks like the MCU is reporting once every ten minutes. Is that to be expected? :)

**Comment by ortegafernando** [comment]
Hi, have you found how to send to the mcu the adjustable max current for the car ? I supposed that orginally your charger could be configured in tuya app, couldnt it ? Thanks.

**Reply by stundenblume** [reply]
> Hi, have you found how to send to the mcu the adjustable max current for the car ? I supposed that orginally your charger could be configured in tuya app, couldnt it ? Thanks.

It's my fathers charger. I will ask him if he did that.

**Reply by ortegafernando** [reply]
@stundenblume  thanks a lot. I hope your answer. Thanks.

**Reply by ortegafernando** [reply]
> > Hi, have you found how to send to the mcu the adjustable max current for the car ? I supposed that orginally your charger could be configured in tuya app, couldnt it ? Thanks.
> 
> It's my fathers charger. I will ask him if he did that.

Hi, sorry, any news? Thanks.

**Reply by stundenblume** [reply]
> Hi, sorry, any news? Thanks.

Hello,
No, nothing new. The charger is hardly ever used, only when an EV comes to visit. Besides, I haven't managed to regulate the power output yet. That was my goal. However, I have a few other projects and would first have to familiarize myself with the charger again. My parents want to buy an EV themselves, so if solar yields increase again, I'll have another reason to look into it.

**Reply by ortegafernando** [reply]
> > Hi, sorry, any news? Thanks.
> 
> Hello, No, nothing new. The charger is hardly ever used, only when an EV comes to visit. Besides, I haven't managed to regulate the power output yet. That was my goal. However, I have a few other projects and would first have to familiarize myself with the charger again. My parents want to buy an EV themselves, so if solar yields increase again, I'll have another reason to look into it.

Thanks a lot, I’ll keep an eye on it. You’ll probably be able to find the Tuya command to control it, won’t you?
